scrypto/src/resource/proof.rs

```diff
@@ -14,23 +14,46 @@ use crate::rust::vec::Vec;
 use crate::rust::string::ToString;
 use crate::types::*;
 
+#[derive(Debug, TypeId, Encode, Decode)]
+pub enum ConsumingProofMethod {
+    Drop(),
+}
+
+#[derive(Debug, TypeId, Encode, Decode)]
+pub enum ProofMethod {
+    Amount(),
+    ResourceAddress(),
+    NonFungibleIds(),
+    Clone(),
+}
+
 /// Represents a proof of owning some resource.
 #[derive(Debug, PartialEq, Eq, Hash)]
 pub struct Proof(pub ProofId);
 
 impl Clone for Proof {
-    fn clone(&self) -> Self {
-        let input = InvokeSNodeInput {
-            snode_ref: SNodeRef::ProofRef(self.0),
-            function: "clone".to_string(),
-            args: args![],
-        };
-        let output: InvokeSNodeOutput = call_engine(INVOKE_SNODE, input);
-        scrypto_decode(&output.rtn).unwrap()
+    invocations! {
+        SNodeRef::ProofRef(self.0) => {
+            fn clone(&self) -> Self { ProofMethod::Clone() }
+        }
     }
 }
 
 impl Proof {
+    invocations! {
+        SNodeRef::ProofRef(self.0) => {
+            pub fn amount(&self) -> Decimal { ProofMethod::Amount() }
+            pub fn resource_address(&self) -> ResourceAddress { ProofMethod::ResourceAddress() }
+            pub fn non_fungible_ids(&self) -> BTreeSet<NonFungibleId> { ProofMethod::NonFungibleIds() }
```

Comment:
How are we going to deal with documentation on the individual methods?

Reply:
We would be able to have better documentation once we have a code generator for these which I believe is the right approach.

Reply:
Before that, can we use the "internal" approach you've been using for the other types?

Reply:
what's the internal approach?